### PR TITLE
Issue 11309: improved check for wanted posts

### LINKED ISSUE #11309: Refuse Diaspora posts no local users actually asked for
I just visited my Community page on my single-user node. I expected it to be a copy of my Network page but I found out it's full of Diaspora posts I've never seen from people I do not follow (otherwise they would be in my timeline).  I assume my node received these posts because of the inverted publication model of Diaspora where users choose who to send their post on top of receiving follow requests.

However, if no local user is actually following the author of a post that was sent to a Friendica node, I believe it should be rejected instead of being stored and displayed in the community page.

### PR DESCRIPTION
**Pull request by annando**
Fixes #11309 by introducing two different levels that indicate why a post had been fetched. Also there is now a similar check for ActivityPub. And finally a lot of logging had been added or improved to improve the support.

### CODE REVIEW

File: src/Protocol/ActivityPub/Processor.php [hunk 1]
```diff
@@ -898,7 +901,7 @@ private static function postMail($activity, $item)
 	 * @return string fetched message URL
 	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
 	 */
-	public static function fetchMissingActivity(string $url, array $child = [], string $relay_actor = '')
+	public static function fetchMissingActivity(string $url, array $child = [], string $relay_actor = '', int $completion = Receiver::COMPLETION_MANUAL)
```
Comment (MrPetovan) [inline src/Protocol/ActivityPub/Processor.php:905]: Please add the DocBlock entry for this new parameter including a reference to the constants describing its possible values.

File: src/Protocol/Diaspora.php [hunk 2]
```diff
@@ -1441,17 +1445,17 @@ private static function storeMentions(int $uriid, string $text)
...
-	 * @param array  $importer Array of the importer user
-	 * @param string $sender   The sender of the message
-	 * @param object $data     The message object
-	 * @param string $xml      The original XML of the message
-	 * @param bool   $fetched  The message had been fetched and not pushed
+	 * @param array  $importer  Array of the importer user
+	 * @param string $sender    The sender of the message
+	 * @param object $data      The message object
+	 * @param string $xml       The original XML of the message
+	 * @param int    $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 3]
```diff
@@ -470,24 +474,24 @@ public static function dispatchPublic($msg, bool $fetched = false)
...
 	 * Dispatches the different message types to the different functions
 	 *
-	 * @param array            $importer Array of the importer user
-	 * @param array            $msg      The post that will be dispatched
-	 * @param SimpleXMLElement $fields   SimpleXML object that contains the message
-	 * @param bool             $fetched  The message had been fetched (default "false")
+	 * @param array            $importer  Array of the importer user
+	 * @param array            $msg       The post that will be dispatched
+	 * @param SimpleXMLElement $fields    SimpleXML object that contains the message
+	 * @param int              $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 4]
```diff
@@ -449,14 +453,14 @@ public static function decode(string $xml, string $privKey = '')
...
-	 * @param bool  $fetched The message had been fetched (default "false")
+	 * @param array $msg       The post that will be dispatched
+	 * @param int   $direction Indicates if the message had been fetched or pushed
 	 *
 	 * @return int The message id of the generated message, "true" or "false" if there was an error
 	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
 	 * @throws \ImagickException
 	 */
-	public static function dispatchPublic($msg, bool $fetched = false)
+	public static function dispatchPublic($msg, int $direction)
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php:463]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 5]
```diff
@@ -1698,15 +1702,16 @@ private static function receiveConversation(array $importer, $msg, $data)
 	/**
 	 * Processes "like" messages
 	 *
-	 * @param array  $importer Array of the importer user
-	 * @param string $sender   The sender of the message
-	 * @param object $data     The message object
+	 * @param array  $importer  Array of the importer user
+	 * @param string $sender    The sender of the message
+	 * @param object $data      The message object
+	 * @param int    $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 6]
```diff
@@ -1885,14 +1890,15 @@ private static function receiveMessage(array $importer, $data)
 	/**
 	 * Processes participations - unsupported by now
 	 *
-	 * @param array  $importer Array of the importer user
-	 * @param object $data     The message object
+	 * @param array  $importer  Array of the importer user
+	 * @param object $data      The message object
+	 * @param int    $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 7]
```diff
@@ -2376,15 +2382,16 @@ private static function addReshareActivity($item, $parent_message_id, $guid, $au
 	/**
 	 * Processes a reshare message
 	 *
-	 * @param array  $importer Array of the importer user
-	 * @param object $data     The message object
-	 * @param string $xml      The original XML of the message
+	 * @param array  $importer  Array of the importer user
+	 * @param object $data      The message object
+	 * @param string $xml       The original XML of the message
+	 * @param int    $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 8]
```diff
@@ -2609,23 +2616,34 @@ private static function receiveRetraction(array $importer, $sender, $data)
 	/**
 	 * Checks if an incoming message is wanted
 	 *
-	 * @param string $url
-	 * @param integer $uriid
+	 * @param array  $item
 	 * @param string $author
 	 * @param string $body
+	 * @param int    $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

File: src/Protocol/Diaspora.php [hunk 9]
```diff
@@ -2651,15 +2669,16 @@ private static function storePhotoAsMedia(int $uriid, $photo)
...
 	 * Receives status messages
 	 *
-	 * @param array            $importer Array of the importer user
-	 * @param SimpleXMLElement $data     The message object
-	 * @param string           $xml      The original XML of the message
-	 * @param bool             $fetched  The message had been fetched and not pushed
+	 * @param array            $importer  Array of the importer user
+	 * @param SimpleXMLElement $data      The message object
+	 * @param string           $xml       The original XML of the message
+	 * @param int              $direction Indicates if the message had been fetched or pushed
```
Comment (MrPetovan) [inline src/Protocol/Diaspora.php]: Please add a reference to the constants the parameter is taking its values from.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Protocol/Diaspora.php (hunk 3 above)
Comment (annando) [reply src/Protocol/Diaspora.php]: How do I do this the best?

Comment (MrPetovan) [reply src/Protocol/Diaspora.php]: You can add something like this: `One of: self::FETCHED, self::PUSHED or self::FORCED_FETCH`

File: src/Protocol/ActivityPub/Processor.php (hunk 1 above)
Comment (MrPetovan) [reply src/Protocol/ActivityPub/Processor.php:905]: Suggestion: `One of the Receiver::COMPLETION_* constant values`

**Comment by MrPetovan** [comment]
There's one docblock request left in `ActivityPub/Processor.php`.

File: src/Protocol/ActivityPub/Processor.php (hunk 1 above)
Comment (annando) [reply src/Protocol/ActivityPub/Processor.php:905]: Done

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**